File: Samples/Layout/ComplexLayout/VB.NET/TestData.vb
Namespace BitMiracle.Docotic.Pdf.Samples
    NotInheritable Class TestData
        Public Shared CreatedOn = New Date(2023, 2, 18, 14, 49, 56, DateTimeKind.Utc)
        Public Shared CompanyName = "Bit Miracle"
        Public Shared Phone = "(000) 111-1111"
        Public Shared Email = "support@bitmiracle.com"

        Public Shared Patient = New PatientInfo With {.Name = "Bruce Wayne", .BirthDate = New Date(1915, 4, 17), .Gender = "M"}
        Public Shared Clinic = New ClinicInfo With {.Name = "Arkham Asylum", .Address = "Gotham City"}
        Public Shared Analysis = New AnalysisInfo With {.Id = Guid.NewGuid().ToString(), .AnalysisDate = Date.UtcNow, .Comment = "Lorem ipsum"}


        Public Shared Results = New ResultItem() {
            New ResultItem("Red Blood Cells", "10e12/L", New ReferenceInterval(4.2, 5.4), 3),
            New ResultItem("Hemoglobin", "g/L", New ReferenceInterval(120, 160), 100.7),
            New ResultItem("Hematocrit", "F of T", New ReferenceInterval(0.37, 0.47), 0.36),
            New ResultItem("MCV", "fL", New ReferenceInterval(78, 100), 90),
            New ResultItem("MCH", "pg", New ReferenceInterval(27, 34), 27),
            New ResultItem("RDW-CV", "%", New ReferenceInterval(11.5, 14), 15.3),
            New ResultItem("Lymphocytes/Neutrophils", Nothing, New ReferenceInterval(0.25, 0.99), 0.15),
            New ResultItem("Lymphocytes %", "F of T", New ReferenceInterval(0.2, 0.75), 0.1)
        }
    End Class

    NotInheritable Class PatientInfo
        Public Name As String
        Public BirthDate As Date
        Public Gender As String
    End Class

    NotInheritable Class ClinicInfo
        Public Name As String
        Public Address As String
    End Class

    NotInheritable Class AnalysisInfo
        Public Id As String
        Public AnalysisDate As Date
        Public Comment As String
    End Class

    NotInheritable Class ResultItem
        Public Sub New(analyte As String, units As String, reference As ReferenceInterval?, result As Double?)
            Me.Analyte = analyte
            Me.Units = units
            Me.Reference = reference
            Me.Result = result
        End Sub

        Public Analyte As String
        Public Units As String
        Public Reference As ReferenceInterval?
        Public Result As Double?

        Public Function FitsReferenceInterval() As ReferenceMatchResult?
            If Reference Is Nothing Or Result Is Nothing Then Return Nothing

            Dim r = Reference.Value
            If Result < r.Min Then Return ReferenceMatchResult.Less

            If Result > r.Max Then Return ReferenceMatchResult.Greater

            Return ReferenceMatchResult.Fit
        End Function
    End Class

    Structure ReferenceInterval
        Public Sub New(min As Double, max As Double)
            Me.Min = min
            Me.Max = max
        End Sub

        Public Min As Double
        Public Max As Double
    End Structure

    Enum ReferenceMatchResult
        Less
        Fit
        Greater
    End Enum
End Namespace

File: Samples/Layout/ComplexLayout/C#/TestData.cs
using System;

namespace BitMiracle.Docotic.Pdf.Samples
{
    static class TestData
    {
        public static DateTime CreatedOn = new(2023, 2, 18, 14, 49, 56, DateTimeKind.Utc);
        public static string CompanyName = "Bit Miracle";
        public static string Phone = "(000) 111-1111";
        public static string Email = "support@bitmiracle.com";

        public static PatientInfo Patient = new("Bruce Wayne", new(1915, 4, 17), "M");
        public static ClinicInfo Clinic = new("Arkham Asylum", "Gotham City");
        public static AnalysisInfo Analysis = new(Guid.NewGuid().ToString(), DateTime.UtcNow, "Lorem ipsum");

        public static ResultItem[] Results = new ResultItem[]
        {
            new("Red Blood Cells", "10e12/L", (4.2, 5.4), 3),
            new("Hemoglobin", "g/L", (120, 160), 100.7),
            new("Hematocrit", "F of T", (0.370, 0.470), 0.36),
            new("MCV", "fL", (78, 100), 90),
            new("MCH", "pg", (27, 34), 27),
            new("RDW-CV", "%", (11.5, 14), 15.3),
            new("Lymphocytes/Neutrophils", null, (0.25, 0.99), 0.15),
            new("Lymphocytes %", "F of T", (0.20, 0.75), 0.1),
        };
    }

    record PatientInfo(string Name, DateTime BirthDate, string Gender);

    record ClinicInfo(string Name, string Address);

    record AnalysisInfo(string Id, DateTime Date, string Comment);

    record ResultItem(
        string Analyte,
        string? Units,
        (double Min, double Max)? Reference,
        double? Result)
    {
        public ReferenceMatchResult? FitsReferenceInterval()
        {
            if (Reference == null || Result == null)
                return null;

            var (min, max) = Reference.Value;
            if (Result < min)
                return ReferenceMatchResult.Less;

            if (Result > max)
                return ReferenceMatchResult.Greater;

            return ReferenceMatchResult.Fit;
        }
    }

    enum ReferenceMatchResult
    {
        Less,
        Fit,
        Greater,
    }
}
